controllers/state_of_the_world.go

```diff
@@ -64,14 +72,124 @@ func NewPolicyMachineryController(manager ctrlruntime.Manager, client *dynamic.D
 
 func buildReconciler(client *dynamic.DynamicClient) controller.ReconcileFunc {
 	reconciler := &controller.Workflow{
-		Precondition: NewEventLogger().Log,
+		Precondition: (&controller.Workflow{
+			Precondition: NewEventLogger().Log,
+			Tasks: []controller.ReconcileFunc{
+				NewTopologyFileReconciler(client, operatorNamespace).Reconcile,
+			},
+		}).Run,
 		Tasks: []controller.ReconcileFunc{
-			NewTopologyFileReconciler(client, operatorNamespace).Reconcile,
+			NewAuthorinoCrReconciler(client).Subscription().Reconcile,
 		},
 	}
 	return reconciler.Run
 }
 
+type AuthorinoCrReconciler struct {
+	Client *dynamic.DynamicClient
+}
+
+func NewAuthorinoCrReconciler(client *dynamic.DynamicClient) *AuthorinoCrReconciler {
+	return &AuthorinoCrReconciler{Client: client}
+}
+
+func (r *AuthorinoCrReconciler) Subscription() *controller.Subscription {
+	return &controller.Subscription{
+		ReconcileFunc: r.Reconcile,
+		Events: []controller.ResourceEventMatcher{
+			{Kind: ptr.To(kuadrantv1beta1.KuadrantKind), EventType: ptr.To(controller.CreateEvent)},
+			{Kind: ptr.To(kuadrantv1beta1.AuthorinoKind), EventType: ptr.To(controller.DeleteEvent)},
+		},
+	}
+}
+
+func (r *AuthorinoCrReconciler) Reconcile(ctx context.Context, _ []controller.ResourceEvent, topology *machinery.Topology, _ error) {
+	logger := controller.LoggerFromContext(ctx).WithName("AuthorinoCrReconciler")
+	logger.Info("reconciling authorino resource", "status", "started")
+	defer logger.Info("reconciling authorino resource", "status", "completed")
+
+	kobjs := lo.FilterMap(topology.Objects().Roots(), func(item machinery.Object, _ int) (*kuadrantv1beta1.Kuadrant, bool) {
+		if item.GroupVersionKind().Kind == kuadrantv1beta1.KuadrantKind.Kind {
+			return item.(*kuadrantv1beta1.Kuadrant), true
+		}
+		return nil, false
+	})
+	if len(kobjs) == 0 {
+		logger.Info("no kuadrant resources found", "status", "skipping")
+		return
+	}
+	if len(kobjs) > 1 {
+		logger.Error(fmt.Errorf("multiple Kuadrant resources found"), "cannot select root Kuadrant resource", "status", "error")
+	}
+	kobj := kobjs[0]
+
+	if kobj.GetDeletionTimestamp() != nil {
+		logger.Info("root kuadrant marked for deletion", "status", "skipping")
+		return
+	}
+
+	aobjs := lo.FilterMap(topology.Objects().Objects().Items(), func(item machinery.Object, _ int) (machinery.Object, bool) {
+		if item.GroupVersionKind().Kind == kuadrantv1beta1.AuthorinoKind.Kind {
+			return item, true
+		}
+		return nil, false
+	})
+
+	if len(aobjs) > 0 {
+		logger.Info("authorino resource already exists, no need to create", "status", "skipping")
+		return
+	}
+
+	authorino := &authorinov1beta1.Authorino{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Authorino",
+			APIVersion: "operator.authorino.kuadrant.io/v1beta1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "authorino",
+			Namespace: kobj.Namespace,
+			OwnerReferences: []metav1.OwnerReference{
```

Comment:
There is a method to add ownerrefs. Up to you.

Reply:
Ya, I looked at that and the ownerrefs method I found at least expects a schema to also be passed in, which we do have access to at this stage. So I went with manually adding the reference.